File: src/karkinolution/decisions/perception.py
from __future__ import annotations

from dataclasses import dataclass, replace
from math import sqrt
from typing import Callable, Iterable

from karkinolution.core.coord import Coord
from karkinolution.core.error import CoordinateNotFoundError

from karkinolution.decisions.actions import MoveActions

from karkinolution.organism.creatures import (
    Corpse,
    Creature,
    EntitiesRegistry,
)
from karkinolution.organism.genetics import CreatureTypes
from karkinolution.organism.identity import EntityTypes, Id
from karkinolution.organism.ontology import Gender, Temperament
from karkinolution.organism.stats import Energy, LimitedValue

from karkinolution.terrain.cell import (
    Cell,
    Damage,
    FoodState,
    MovimentCost,
    Properties,
)
from karkinolution.terrain.map import (
    EntityMap,
    Geometry,
    Territory,
)

@dataclass(frozen=True)
class PerceivedBody:
    energy:Energy
    life:int | float
    physical_ratio:float
    reproductiv_capacity:bool

@dataclass(frozen=True)
class PerceivedOntology:
    specie:CreatureTypes
    temperament:Temperament
    gender:Gender | None

@dataclass(frozen=True)
class PerceivedCreature:
    body:PerceivedBody
    ontology:PerceivedOntology
    id:Id

@dataclass(frozen=True)
class PerceivedCorpse:
    energy:Energy
    id:Id

@dataclass(frozen=True)
class PerceivedCell:
    is_movable:bool
    is_edible:bool
    is_dangerous:bool
    
    movement_cost:int | float | None
    food:Energy | None
    required_capabilities: set[MoveActions]
    damage:int | float | None

@dataclass(frozen=True)
class ObserverCreature:
    energy_ratio:float
    life_ratio:float
    id:Id
    specie:CreatureTypes

@dataclass(frozen=True)
class PerceivedBlock:
    cell:PerceivedCell
    entity:PerceivedCreature | PerceivedCorpse | None
    distance:float

    @property
    def has_entity(self) -> bool:
        return self.entity is not None
    
    
    def get_entity_type(self) -> EntityTypes | None:
        return None if not self.has_entity else self.entity.id.e_type # type: ignore
    @property
    def has_creature(self) -> bool:
        return self.get_entity_type() == EntityTypes.CREATURE
        
    @property
    def has_corpse(self) -> bool:
        return self.get_entity_type() == EntityTypes.CORPSE
    # Alias
    # Creature = PerceivedCreature
    # Corpse = PerceivedCorpse

    # has_entity = true -> isinstance(entity, Creature) = true | isinstance(entity, Corpse) = true
    # has_creature = true -> isinstance(entity, Creature) = true
    # has_corpse = true -> isinstance(entity, Corpse) = true

    
class BlockProperties:
    @staticmethod
    def has_aggressive_creature(block:PerceivedBlock) -> bool:
        return block.has_creature and block.entity.ontology.temperament == Temperament.AGGRESSIVE # type: ignore
    @staticmethod
    def has_territorial_creature(block:PerceivedBlock) -> bool:
        return block.has_creature and block.entity.ontology.temperament == Temperament.TERRITORIAL # type: ignore
    @staticmethod
    def is_same_specie(other_creature:PerceivedCreature, creature:ObserverCreature) -> bool:
        return other_creature.specie == creature.specie  # type: ignore
    @staticmethod
    def is_different_specie(other:PerceivedCreature, creature:ObserverCreature) -> bool:
        return other.specie != creature.specie # type: ignore
    @staticmethod
    def is_more_strong(other:PerceivedCreature, creature:Creature) -> bool:
        return other.body.physical_ratio > creature.physical_ratio
    
@dataclass(frozen=True)
class Perception:
    pieces:dict[Coord, PerceivedBlock]
    creature:ObserverCreature
    coord:Coord
    max_distance:float

    def __iter__(self) -> Iterable[Coord]:
        return iter(self.pieces.keys())

    @property
    def iter(self) -> Iterable[tuple[Coord, PerceivedBlock]]:
        return self.pieces.items()
    @property
    def blocks(self) -> Iterable[PerceivedBlock]:
        return self.pieces.values()
    @property
    def coords(self) -> Iterable[Coord]:
        return self.pieces.keys()
    @property
    def size(self) -> int:
        return len(self.pieces)
    
    
    def get(self, coord:Coord) -> PerceivedBlock:
        try:
            return self.pieces[coord]
        except KeyError:
            raise CoordinateNotFoundError("coord {} was not found".format(coord))
    def try_get(self, coord:Coord) -> PerceivedBlock | None:
        return self.pieces.get(coord)
    @property
    def creature_block(self) -> PerceivedBlock:
        return self.get(self.coord)


class PerceptionAnalyser:
    @staticmethod
    def build_another_perception(old_perception:Perception, new_pieces:dict[Coord, PerceivedBlock]) -> Perception:
        return replace(old_perception, pieces=new_pieces)
    @staticmethod
    def find_predicate(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool]) -> Perception:
        pieces:dict[Coord, PerceivedBlock] = {}

        for coord, block in perception.iter:
            if predicate(block, coord):
                pieces[coord] = block
        return PerceptionAnalyser.build_another_perception(perception, pieces)
    @staticmethod
    def find_predicate_coords(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool]) -> set[Coord]:
        pieces:set[Coord] = set()

        for coord, block in perception.iter:
            if predicate(block, coord):
                pieces.add(coord)
        return pieces
    @staticmethod
    def build(perception:Perception, coords:set[Coord] | list[Coord] | tuple[Coord], predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        pieces:dict[Coord, PerceivedBlock] = {}

        for coord in coords:
            perceived_block = perception.try_get(coord)
            
            if perceived_block is not None and predicate(perceived_block, coord):
                pieces[coord] = perceived_block
        return PerceptionAnalyser.build_another_perception(perception, pieces)
    
    @staticmethod
    def neighbors_4(perception:Perception, include_self_coord:bool = False) -> Perception:
        coords = perception.coord.four_movements()

        if include_self_coord:
            coords.add(perception.coord)
        return PerceptionAnalyser.build(perception, coords)
    
    @staticmethod
    def neighbors_8(perception:Perception, include_self_coord:bool = False) -> Perception:
        coords = perception.coord.eight_movements()

        if include_self_coord:
            coords.add(perception.coord)
        return PerceptionAnalyser.build(perception, coords)

    @staticmethod
    def neighbors_x_y(perception:Perception, x_radius:int, y_radius:int) -> Perception: 
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda _, c:  c.x - perception.coord.x > x_radius or c.y - perception.coord.y > y_radius or c == perception.coord)
    @staticmethod
    def near_coord(coord_creature:Coord, coords:set[Coord] | list[Coord] | tuple[Coord]) -> Coord:
        return min(coords, key=lambda x: coord_creature.distance_to_other(x))
    @staticmethod
    def get_area_in_radius_ratio(perception:Perception, radius_ratio:float) -> Perception:
        assert 0 <= radius_ratio<= 1
        return PerceptionAnalyser.find_predicate(perception, lambda _, c: c.distance_to_other(perception.coord) < radius_ratio * perception.max_distance )
    
    
class PerceptionPatterns:
    @staticmethod
    def empty_spaces(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda b, c: not b.has_entity and predicate(b, c))
    @staticmethod
    def corpses(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda b, c: b.has_corpse and predicate(b, c))
    
    @staticmethod
    def creatures(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda b, c: b.has_creature and predicate(b, c))
    @staticmethod
    def same_species(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda b, c: b.has_creature and b.entity.ontology.specie == perception.creature.specie and predicate(b, c)) # type: ignore
    @staticmethod
    def other_species(perception:Perception, predicate:Callable[[PerceivedBlock, Coord], bool] = lambda b, c: True) -> Perception:
        return PerceptionAnalyser.find_predicate(perception, predicate=lambda b, c: b.has_creature and b.entity.ontology.specie != perception.creature.specie and predicate(b, c)) # type: ignore


class Perceiver:
    @staticmethod
    def perceive_entity(creature:Creature) -> PerceivedCreature:
        body = PerceivedBody(
            creature.energy,
            creature.life.value,
            creature.physical_ratio,
            creature.reproductively_capable
        )
        ontology = PerceivedOntology(
            creature.genome.core.specie,
            creature.genome.core.temperament,
            creature.gender
        )

        return PerceivedCreature(
            body,
            ontology,
            creature.id
        )
    @staticmethod
    def perceive_corpse(corpse:Corpse) -> PerceivedCorpse:
        return PerceivedCorpse(
            corpse.energy,
            corpse.id
        )
    
    @staticmethod
    def perceive_cell(cell:Cell) -> PerceivedCell:
        is_edible = cell.property_is_in(Properties.EDIBLE)

        food: None | Energy = None
        if is_edible:
            food = cell.get_component(FoodState).food # type: ignore
        
        is_movable = True if len(cell.required_capabilities) > 0 else False

        movement_cost = None
        if is_movable:
            movement_cost = cell.get_component(MovimentCost).moviment_cost # type: ignore
        
        is_dangerous = cell.property_is_in(Properties.DANGEROUS)
        damage = None
        if is_dangerous:
            damage = cell.get_component(Damage).damage # type: ignore
        
        return PerceivedCell(
            is_movable,
            is_edible,
            cell.property_is_in(Properties.DANGEROUS),
            movement_cost,
            food,
            cell.required_capabilities,
            damage
        )
            
def perceive(creature:Creature, territory:Territory, entity_map:EntityMap,
             entities:EntitiesRegistry) -> Perception:
    vision_radius = creature.genome.core.vision_radius

    creature_observer = ObserverCreature(
        creature.energy.ratio,
        creature.life.ratio,
        creature.id,
        creature.genome.core.id,
    )
    data = Geometry.neighbors_x_y(creature.position, territory, entity_map, vision_radius, True)

    perceived:dict[Coord, PerceivedBlock] = {}

    for c, b in data.items():
        perceived_cell = Perceiver.perceive_cell(b.cell)

        id = b.id_creature

        perceived_creature = None
        
        if id is None:
            pass
        elif id.e_type == EntityTypes.CREATURE:
            perceived_creature = Perceiver.perceive_entity(entities.get_creature(id))
        elif id.e_type == EntityTypes.CORPSE:
            perceived_creature = Perceiver.perceive_corpse(entities.get_corpse(id))
        

        perceived[c] = PerceivedBlock(perceived_cell, perceived_creature, creature.position.distance_to_other(c))

    return Perception(perceived, creature_observer, creature.position, sqrt(creature.genome.core.vision_radius.x**2 + creature.genome.core.vision_radius.y**2))

@dataclass
class Dangers:
    cell_danger:float
    creature_danger:float

    @property
    def accumulated_danger(self) -> float:
        return (self.cell_danger * 0.85 + self.creature_danger * 1.15)/2
    


TERRITORIAL_DANGER = 0.25
AGGRESSIVE_DANGER = 0.3
DANGEROUS_CELL_FACTOR = 0.18
ADJACENT_DANGER_WEIGHT = 0.4


@dataclass(frozen=True)
class DangerIndex:
    index:dict[Coord, Dangers]
    
    def get(self, coord:Coord) -> Dangers:
        return self.index[coord]

    def try_get(self, coord:Coord) -> Dangers | None:
        return self.index.get(coord)
    def add(self, coord:Coord, dangers:Dangers) -> None:
        if not coord in self.index:
            self.index[coord] = dangers

class DangerFactory:
    # all values must be in [0, 1]
    @staticmethod
    def get_creature_danger(block:PerceivedBlock, creature:Creature) -> float:
        score = LimitedValue(0, 1, -1)

        if block.has_creature:
            other:PerceivedCreature = block.entity # type: ignore
            if BlockProperties.has_territorial_creature(block):
                score.add(TERRITORIAL_DANGER)
            elif BlockProperties.has_aggressive_creature(block):
                score.add(AGGRESSIVE_DANGER)
        
            score.sub((creature.physical_ratio - other.body.physical_ratio)/2)
        return score.value
    @staticmethod
    def get_cell_danger(block:PerceivedBlock) -> float:
        score = LimitedValue(0, 1, -1)

        if block.cell.is_dangerous:
            score.add(DANGEROUS_CELL_FACTOR)
        return score.value
    @staticmethod
    def get_dangers(block:PerceivedBlock, creature:Creature) -> Dangers:
        return Dangers(
            DangerFactory.get_cell_danger(block),
            DangerFactory.get_creature_danger(block, creature)
        )
    @staticmethod
    def create_danger_index(perception:Perception, creature:Creature) -> DangerIndex:
        cache = DangerIndex({})
        danger_index = DangerIndex({})
        new_perception = PerceptionAnalyser.get_area_in_radius_ratio(perception, 0.75)


        for c, b in new_perception.iter:
            if c in cache.index:
                actual_dangers = cache.get(c)
            else:
                actual_dangers = DangerFactory.get_dangers(b, creature)
                cache.add(c, actual_dangers)
            
            four_coords = c.four_movements()

            average_dangers = Dangers(0, 0)
            count = 0
            for f_c in four_coords:
                if f_c not in perception.pieces:
                    continue

                if f_c not in cache.index:
                    dangers = DangerFactory.get_dangers(perception.get(f_c), creature)
                    cache.add(f_c, dangers)
                else:
                    dangers = cache.get(f_c)
                
                count += 1


                # sums
                average_dangers.cell_danger += dangers.cell_danger
                average_dangers.creature_danger += dangers.creature_danger

            if count > 0:
                average_dangers.cell_danger/=count
                average_dangers.creature_danger/=count

                average_dangers.cell_danger*=ADJACENT_DANGER_WEIGHT
                average_dangers.creature_danger*=ADJACENT_DANGER_WEIGHT

            new_dangers = Dangers(
                average_dangers.cell_danger + actual_dangers.cell_danger,
                average_dangers.creature_danger + actual_dangers.creature_danger
            )
            danger_index.add(c, new_dangers)
        return danger_index

File: src/karkinolution/brain/perception.cpp
#include <karkinolution/terrain/map.hpp>
#include <karkinolution/terrain/soil.hpp>
#include <karkinolution/core/vec2.hpp>
#include <karkinolution/brain/perception.hpp>
#include <karkinolution/organism/ontology.hpp>
#include <optional>
#include <ranges>
#include <karkinolution/core/stats.hpp>
#include <variant>


bool PerceivedBlock::has_entity() const {
    return !std::holds_alternative<std::monostate>(entity);
}

std::optional<EntityTypes> PerceivedBlock::get_entity_type() const {
    if (std::holds_alternative<PerceivedCreature>(entity)) {
        return EntityTypes::CREATURE;
    }
    if (std::holds_alternative<PerceivedCorpse>(entity)) {
        return EntityTypes::CORPSE;  
    }
    return std::nullopt;
}

bool PerceivedBlock::has_creature() const {
    return std::holds_alternative<PerceivedCreature>(entity);
}

bool PerceivedBlock::has_corpse() const {
    return std::holds_alternative<PerceivedCorpse>(entity);
}

const PerceivedCreature& PerceivedBlock::get_creature() const {
    return std::get<PerceivedCreature>(entity);
}

const PerceivedCorpse& PerceivedBlock::get_corpse() const {
    return std::get<PerceivedCorpse>(entity);
}

// BlockPropeties

bool BlockProperties::has_aggressive_creature(const PerceivedBlock& block) {
    if (auto* creature = std::get_if<PerceivedCreature>(&block.entity)) {
        return creature->ontology.temperament == Temperament::AGGRESSIVE;
    }
    return false;
}

bool BlockProperties::has_territorial_creature(const PerceivedBlock& block) {
    if (auto* creature = std::get_if<PerceivedCreature>(&block.entity)) {
        return creature->ontology.temperament == Temperament::TERRITORIAL;
    }
    return false;
}

bool BlockProperties::is_same_specie(const PerceivedCreature& other, const ObserverCreature& creature) {
    return other.ontology.specie == creature.specie;
}

bool BlockProperties::is_more_strong(const PerceivedCreature& other, const ObserverCreature& creature) {
    return other.body.physical_ratio > creature.life_ratio;
}


// Perception

const_iterator Perception::begin() const {
    return pieces.begin();
}

const_iterator Perception::end() const {
    return pieces.end();
}

int Perception::size() const {
    return static_cast<int>(pieces.size());
}

const std::unordered_map<Vec2, PerceivedBlock>& Perception::pieces_ref() const {
    return pieces;
}

bool Perception::contains(Vec2 coord) const {
    return pieces.contains(coord);
}

bool Perception::empty() const {
    return pieces.empty();
}

const PerceivedBlock& Perception::at(Vec2 coord) const {
    return pieces.at(coord);
}

const PerceivedBlock* Perception::try_at(Vec2 coord) const {
    auto it = pieces.find(coord);
    if (it != pieces.end()) {
        return &(it->second); 
    }
    return nullptr;
}


const PerceivedBlock& Perception::creature_block() const {
    return pieces.at(coord());
}

const ObserverCreature& Perception::creature() const {
    return _creature;
}

Vec2 Perception::coord() const {
    return _coord;
}

float Perception::max_distance() const {
    return _max_distance;
}

// PerceptionAnalyser
Perception PerceptionAnalyser::build_another_perception(
    const Perception& old_perception,
    const std::unordered_map<Vec2, PerceivedBlock>& new_pieces) 
{
    Perception new_p{
        new_pieces,
        old_perception.creature(),
        old_perception.coord(),
        old_perception.max_distance()
    };
    return new_p;
}
    

Perception PerceptionAnalyser::neighbors_4(const Perception& perception, bool include_self) {
    return PerceptionAnalyser::build(perception, Vec2F::four_movements(perception.coord()));
}

Perception PerceptionAnalyser::neighbors_8(const Perception& perception, bool include_self) {
    return PerceptionAnalyser::build(perception, Vec2F::eight_movements(perception.coord()));
}


Perception PerceptionAnalyser::get_area_in_radius_ratio(const Perception& perception, float radius_ratio) {
    std::unordered_map<Vec2, PerceivedBlock> filtered;
    float current_max_search = perception.max_distance() * radius_ratio;

    for (const auto& [pos, block] : perception) {
        if (block.distance <= current_max_search) {
            filtered[pos] = block;
        }
    }
    return build_another_perception(perception, filtered);
}

Vec2 PerceptionAnalyser::near_coord(const Perception&perception) {
    auto size = perception.size();
    Vec2 near_coord = perception.begin()->first;
    float near_distance = perception.begin()->second.distance;
    
        for (auto i = 0; i < size + 1; i++) {
            const PerceivedBlock* block = &(perception.begin()++)->second;
            const float distance = block->distance;
            if (distance < near_distance) {
                near_coord = (perception.begin()++)->first;
                near_distance = distance;
            }
        }
    return near_coord;
}


// Dangers, DangerIndex, DangerFactory

float Dangers::accumulated_danger() const {
    return cell_danger.value() + creature_danger.value();
}

const Dangers& DangerIndex::at(Vec2 coord) const {
    return index.at(coord);
}

const Dangers* DangerIndex::try_at(Vec2 coord) const {
    auto it = index.find(coord);
    if (it != index.end()) {
        return &(it->second);
    }
    return nullptr;
}

void DangerIndex::add(Vec2 coord, const Dangers& dangers) {
    if (try_at(coord) == nullptr) {
        index[coord] = dangers;
    } 
}

// Perceiver

PerceivedCreature Perceiver::perceive_creature(const Creature&creature) {
    PerceivedBody body{
        .energy = creature.energy,
        .life = creature.life.value(),
        .physical_ratio = creature.physical_ratio(),
        .reproductive_capacity = creature.reproductively_capable()
    };

    PerceivedOntology ontology{
        .specie = creature.genome.core.specie,
        .temperament = creature.genome.core.temperament,
        .gender = creature.gender
    };

    return PerceivedCreature{
        .body = body,
        .ontology = ontology,
        .id = IDF::create_creature_id(creature.id)
    };
}

PerceivedCorpse Perceiver::perceive_corpse(const Corpse &corpse) {
    return PerceivedCorpse{
        .energy = corpse.energy,
        .id = IDF::create_corpse_id(corpse.id)
    };
}

PerceivedSoil Perceiver::perceive_soil(const SoilPiece &soil_piece) {
    bool is_movable = false;
    bool is_edible = false;
    bool is_dangerous = false;

    for (auto& property : soil_piece.properties) {
        if (property == Properties::DANGEROUS) {
            is_dangerous = true;
        } else if (property == Properties::EDIBLE) {
            is_edible = true;
        }
    }

    if (soil_piece.components.exists<MovementCost>()) { is_movable = true;}

    std::optional<float> movement_cost = std::nullopt, damage = std::nullopt;
    std::optional<Energy> food = std::nullopt;

    if (is_dangerous) {
        damage = soil_piece.components.try_get<Damage>()->damage;
    }

    if (is_edible) {
        food = soil_piece.components.try_get<FoodState>()->food;
    }
    if (is_movable) {
        movement_cost = soil_piece.components.try_get<MovementCost>()->cost;
    }
    return PerceivedSoil{
        .is_movable = is_movable,
        .is_edible = is_edible,
        .is_dangerous = is_dangerous,
        .movement_cost = movement_cost,
        .damage = damage,
        .food = food
    };
}


Perception Perceiver::perceive(const Creature &creature, const Territory &territory, const EntityMap &entity_map, const EntitiesRegistry &entities) {
    const Vec2 vision_radius = creature.genome.core.vision_radius;

    ObserverCreature observer_creature{
        .energy_ratio = creature.energy.ratio(),
        .life_ratio = creature.life.ratio(),
        .specie = creature.genome.core.specie,
        .id = IDF::create_creature_id(creature.id),
    };

    std::unordered_map<Vec2, PerceivedBlock> perceived{};
    perceived.reserve((2*creature.genome.core.vision_radius.x + 1) * (2*creature.genome.core.vision_radius.y + 1));

    for (int x = -creature.genome.core.vision_radius.x; x <= creature.genome.core.vision_radius.x; x++) {
        for (int y = -creature.genome.core.vision_radius.y; y <= creature.genome.core.vision_radius.y; y++) {
            const Vec2 coord{x + creature.position.x, y + creature.position.y};
            if (territory.exists(coord)) {
                PerceivedSoil perceived_soil = Perceiver::perceive_soil(territory.at(coord));
                
                const Id* id = entity_map.try_at(coord);
                const Entity* entity = nullptr;
                PerceivedEntity perceived_entity = std::monostate();
                if (id) {
                    entity = &entities.at(*id);
                    if (id->entity_type == EntityTypes::CREATURE) {
                        perceived_entity = PerceivedCreature{Perceiver::perceive_creature(std::get<Creature>(*entity))};
                    } else if (id->entity_type == EntityTypes::CORPSE) {
                        perceived_entity = PerceivedCorpse{Perceiver::perceive_corpse(std::get<Corpse>(*entity))};
                    }
                }
                perceived.emplace(coord,
                    PerceivedBlock{.cell=perceived_soil, .entity=perceived_entity, .distance=Vec2F::distance(creature.position, coord)});
            }
        }
    }
    
    return Perception{
    perceived,
    observer_creature,
    creature.position,
    creature.position.length()
    };
}


// DangerFactory

SignedNormalizedValue DangerFactory::get_creature_danger(const PerceivedBlock& block, const Creature& creature) {
    SignedNormalizedValue score{0};

    if (auto* other = std::get_if<PerceivedCreature>(&block.entity)) {
        if (BlockProperties::has_aggressive_creature(block)) {
            score += AGGRESSIVE_DANGER;
        }
        if (BlockProperties::has_territorial_creature(block)) {
            score += TERRITORIAL_DANGER;
        }
        score -= (creature.physical_ratio().value() - other->body.physical_ratio.value())/2;
    }

    return score.value();
}

SignedNormalizedValue DangerFactory::get_cell_danger(const PerceivedBlock& block) {
    SignedNormalizedValue danger = 0.0f;
    if (block.cell.is_dangerous) {
        danger += DANGEROUS_CELL_FACTOR;
    }
    if (block.cell.damage.has_value()) {
        danger += block.cell.damage.value() * 0.01f;
    }
    return danger;
}

Dangers DangerFactory::get_dangers(const PerceivedBlock& block, const Creature& creature) {
    return Dangers {
        .cell_danger = get_cell_danger(block),
        .creature_danger = get_creature_danger(block, creature)
    };
}

DangerIndex DangerFactory::create_danger_index(const Perception& perception, const Creature& creature) {
    DangerIndex danger_index;
    auto new_perception = PerceptionAnalyser::get_area_in_radius_ratio(perception, GET_DANGER_INDEX_AREA_RATIO);


    for (const auto& [coord, block] : perception) {
        Dangers block_dangers = get_dangers(block, creature);
        Dangers four_coords_dangers = Dangers{0, 0};
        std::vector<Vec2> four_movements = Vec2F::four_movements(coord);
        int count = 0; 
        
        for (auto f_c : four_movements)
            if (new_perception.contains(f_c)) {
                four_coords_dangers.cell_danger += DangerFactory::get_cell_danger(block).value();
                four_coords_dangers.creature_danger += DangerFactory::get_creature_danger(block, creature).value();
                count ++;
            }
        if (count > 0) {
            four_coords_dangers.cell_danger/= static_cast<float>(count);
            four_coords_dangers.creature_danger/= static_cast<float>(count);

            four_coords_dangers.cell_danger *= ADJACENT_DANGER_WEIGHT;
            four_coords_dangers.creature_danger *= ADJACENT_DANGER_WEIGHT;
        }

        Dangers main_dangers = {
            .cell_danger = block_dangers.cell_danger.value() + four_coords_dangers.cell_danger.value(),
            .creature_danger = block_dangers.creature_danger.value() + four_coords_dangers.creature_danger.value() 
        };
    }
    return danger_index;
}